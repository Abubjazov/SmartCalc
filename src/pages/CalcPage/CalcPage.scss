@import '../../variables';

.calc-page {
	background: linear-gradient(
		180deg,
		rgba(254, 254, 254, 1) 0%,
		rgba(238, 238, 238, 1) 100%
	);

	.container {
		padding: 11vh 0 1vh 0;
		min-height: 100vh;
		max-width: calc(100vw - 2vw);
		display: grid;
		grid-template-columns: 1fr 1fr;
		grid-auto-rows: min-content;
		row-gap: 10px;
		justify-items: center;

		@media (max-width: 790px) {
			grid-template-columns: 1fr;
			align-items: center;

			.calc-info {
				display: none;
			}
		}

		.calc-info {
			background-color: $white;
			border-radius: 10px;
		}
	}
}
