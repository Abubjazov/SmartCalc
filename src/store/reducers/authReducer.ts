import { AuthAction, AuthActionTypes, AuthState } from '../../interfaces'

const authDataString = localStorage.getItem('smartcalc_token')
const authData = authDataString ? JSON.parse(authDataString) : null

export const initialState: AuthState = {
	token: authData?.token || null,
	name: authData?.name || '',
	email: authData?.email || '',
	status: 'waiting',
	error: null,
}

export const authReducer = (
	state: AuthState = initialState,
	action: AuthAction
): AuthState => {
	switch (action.type) {
		case AuthActionTypes.FETCH_TOKEN:
			return {
				...state,
				status: 'loading',
				error: null,
			}

		case AuthActionTypes.FETCH_TOKEN_SUCCESS:
			return {
				...state,
				status: 'waiting',
				token: action.payload.token,
				name: action.payload.name,
				email: action.payload.email,
			}

		case AuthActionTypes.FETCH_TOKEN_ERROR:
			return {
				...state,
				status: 'error',
				error: action.payload,
			}

		case AuthActionTypes.DELETE_TOKEN:
			return {
				...state,
				status: 'loading',
				error: null,
			}

		case AuthActionTypes.DELETE_TOKEN_SUCCESS:
			return {
				...state,
				status: 'waiting',
				token: null,
				name: '',
				email: '',
			}

		case AuthActionTypes.DELETE_TOKEN_ERROR:
			return {
				...state,
				status: 'error',
				error: action.payload,
			}

		default:
			return state
	}
}
