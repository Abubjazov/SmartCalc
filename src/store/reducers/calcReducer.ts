import { nanoid } from 'nanoid'

import { CalcAction, CalcActionTypes, CalcState } from '../../interfaces'
import { matchCheck, returnInputItems, returnSummary } from '../../utils/utils'

export const initialState: CalcState = {
	items: [],
	inputItems: [
		{ key: nanoid(), value: '' },
		{ key: nanoid(), value: '' },
	],
	sortedItems: [],
	sortDirection: false,
	searchString: '',
	status: 'waiting',
	step: 1,
	summary: 0,
	error: null,
}

export const calcReducer = (
	state: CalcState = initialState,
	action: CalcAction
): CalcState => {
	switch (action.type) {
		case CalcActionTypes.FETCH_CURRENT_STATE:
			return { ...state, status: 'loading', error: null }

		case CalcActionTypes.FETCH_CURRENT_STATE_SUCCESS:
			return {
				...state,
				status: 'waiting',
				items: action.payload.items,
				sortedItems: action.payload.items,
				step: action.payload.step,
				inputItems: returnInputItems(action.payload.items),
				summary: returnSummary(action.payload.items),
			}

		case CalcActionTypes.FETCH_CURRENT_STATE_ERROR:
			return {
				...state,
				status: 'error',
				error: action.payload,
			}

		case CalcActionTypes.SWITCH_TO_CONFIRM:
			return { ...state, status: 'loading', error: null }

		case CalcActionTypes.SWITCH_TO_CONFIRM_SUCCESS:
			return {
				...state,
				status: 'waiting',
				items: action.payload.items,
				sortedItems: action.payload.items,
				step: action.payload.step,
			}

		case CalcActionTypes.SWITCH_TO_CONFIRM_ERROR:
			return {
				...state,
				status: 'error',
				error: action.payload,
			}

		case CalcActionTypes.ADD_INPUT_ITEM:
			return {
				...state,
				inputItems: [...state.inputItems, action.payload],
			}

		case CalcActionTypes.CHANGE_INPUT_ITEM:
			return {
				...state,
				inputItems: state.inputItems.map(item => {
					if (item.key === action.payload.key) {
						item.value = action.payload.value
					}
					return item
				}),
			}

		case CalcActionTypes.REMOVE_INPUT_ITEM:
			return {
				...state,
				inputItems: state.inputItems.filter(
					item => item.key !== action.payload
				),
			}

		case CalcActionTypes.SORT_ITEMS:
			return {
				...state,
				sortedItems: state.sortDirection
					? state.sortedItems.sort((a, b) => a - b)
					: state.sortedItems.sort((a, b) => b - a),
				sortDirection: !state.sortDirection,
			}

		case CalcActionTypes.SEARCH_ITEMS:
			return {
				...state,
				searchString: action.payload,
				sortedItems: state.items.filter(item =>
					matchCheck(item, action.payload)
				),
			}

		case CalcActionTypes.SWITCH_TO_DATAINPUT:
			return { ...state, status: 'loading', error: null }

		case CalcActionTypes.SWITCH_TO_DATAINPUT_SUCCESS:
			return {
				...state,
				status: 'waiting',
				items: action.payload.items,
				searchString: '',
				sortedItems: action.payload.items,
				step: action.payload.step,
				inputItems: [
					{ key: 'a13key1', value: '' },
					{ key: 'a13key2', value: '' },
				],
			}

		case CalcActionTypes.SWITCH_TO_DATAINPUT_ERROR:
			return {
				...state,
				status: 'error',
				error: action.payload,
			}

		case CalcActionTypes.SWITCH_TO_RESULT:
			return { ...state, status: 'loading', error: null }

		case CalcActionTypes.SWITCH_TO_RESULT_SUCCESS:
			return {
				...state,
				status: 'waiting',
				items: action.payload.items,
				searchString: '',
				sortedItems: action.payload.items,
				step: action.payload.step,
				summary: action.payload.summary,
			}

		case CalcActionTypes.SWITCH_TO_RESULT_ERROR:
			return {
				...state,
				status: 'error',
				error: action.payload,
			}

		case CalcActionTypes.RESET_TO_DATA_INPUT:
			return { ...state, status: 'loading', error: null }

		case CalcActionTypes.RESET_TO_DATA_INPUT_SUCCESS:
			return {
				...state,
				status: 'waiting',
				items: action.payload.items,
				sortedItems: action.payload.items,
				step: action.payload.step,
				summary: 0,
			}

		case CalcActionTypes.RESET_TO_DATA_INPUT_ERROR:
			return {
				...state,
				status: 'error',
				error: action.payload,
			}

		case CalcActionTypes.RESET_TO_NEW_CALCULATION:
			return { ...state, status: 'loading', error: null }

		case CalcActionTypes.RESET_TO_NEW_CALCULATION_SUCCESS:
			return {
				...state,
				status: 'waiting',
				items: action.payload.items,
				sortedItems: action.payload.items,
				step: action.payload.step,
				inputItems: [
					{ key: 'a13key1', value: '' },
					{ key: 'a13key2', value: '' },
				],
				summary: 0,
			}

		case CalcActionTypes.RESET_TO_NEW_CALCULATION_ERROR:
			return {
				...state,
				status: 'error',
				error: action.payload,
			}

		default:
			return state
	}
}
