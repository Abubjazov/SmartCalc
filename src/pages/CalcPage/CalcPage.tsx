import { Calculator } from '../../components'

import './CalcPage.scss'

export const CalcPage = (): JSX.Element => (
	<main className='calc-page'>
		{/* <PageMarker pageName='calc' /> */}

		<div className='container'>
			<div className='calc-info'>
				<h2>Smart Calc - простой онлайн калькулятор.</h2>
				<hr />
				<br />
				Smart C - простое онлайн приложение на основе React, Redux и Typescript.
				<br />
				<br />
				<h3>Логин и пароль для Demo:</h3>
				<br />
				login: test@test.com <br /> password: password
				<br />
				<br />
				<h3>На данный момент приложение имеет 3 реализованных раздела:</h3>
				<br />
				-Главная страница - первое что видит пользователь.
				<br />
				<br />
				- Страница ввода логина и пароля - где пользователь уже
				зарегистрированный в системе может ввести свой логин и пароль и попасть
				на страницу "Калькулятор".
				<br />
				<br />- Страница калькулятора - где пользователь может ввести данные и
				получить результат расчёта.
			</div>
			<Calculator />
		</div>
	</main>
)
