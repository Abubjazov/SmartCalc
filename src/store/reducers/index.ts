import { combineReducers } from 'redux'
import { authReducer } from './authReducer'
import { calcReducer } from './calcReducer'

export const rootReducer = combineReducers({
	calc: calcReducer,
	auth: authReducer,
})

export type RootState = ReturnType<typeof rootReducer>
