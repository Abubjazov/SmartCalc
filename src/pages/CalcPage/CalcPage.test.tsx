import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import { BrowserRouter } from 'react-router-dom'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'

import { CalcPage } from './CalcPage'

import { inputItemsInitValue } from '../../store/reducers/calcReducer'

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

describe('Component: CalcPage', () => {
	test('should render CalcPage default', () => {
		const store = mockStore({
			auth: {
				token: 'testtoken',
				email: 'test@test.com',
			},
			calc: {
				step: 1,
				status: 'waiting',
				error: null,
				inputItems: [...inputItemsInitValue],
			},
		})

		const { asFragment } = render(
			<Provider store={store}>
				<BrowserRouter>
					<CalcPage />
				</BrowserRouter>
			</Provider>
		)

		expect(asFragment()).toMatchSnapshot()
	})
})
