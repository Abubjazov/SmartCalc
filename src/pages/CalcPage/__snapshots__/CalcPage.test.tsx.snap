// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: CalcPage should render CalcPage default 1`] = `
<DocumentFragment>
  <main
    class="calc-page"
  >
    <div
      class="container"
    >
      <div
        class="calc-info"
      >
        <h2>
          SMART CALC V.2
        </h2>
        <h3>
          Онлайн калькулятор
        </h3>
        <p>
          SPA на основе React, Redux и Typescript
        </p>
        <p>
          Репозиторий проекта:
        </p>
        <a
          href="https://github.com/Abubjazov/SmartCalc"
          rel="noopener noreferrer"
          target="_blank"
        >
          github.com/Abubjazov/SmartCalc
        </a>
        <p>
          Документация:
        </p>
        <a
          href="https://github.com/Abubjazov/SmartCalc/blob/main/README.md"
          rel="noopener noreferrer"
          target="_blank"
        >
          README.md
        </a>
        <p>
          Demo сервер:
        </p>
        <a
          href="https://abubjazov.github.io/SmartCalc/"
          rel="noopener noreferrer"
          target="_blank"
        >
          github.io/SmartCalc/
        </a>
        <p>
          Стэк технологий:
        </p>
        <h3>
          Typescript, React, ReactRouter, ReactHelmet, Redux, ReduxThunk, Axios, Nanoid, Bcryptjs, SASS
        </h3>
        <p
          class="calc-info-clarification"
        >
          Все права защищены! © Abubjazov 2022 (
          <a
            href="https://github.com/Abubjazov"
            rel="noopener noreferrer"
            target="_blank"
          >
            https://github.com/Abubjazov
          </a>
          )
        </p>
      </div>
      <div
        class="calc-widget"
      >
        <p
          class="calc-widget-p"
        >
          Привет, Test@t***.com
        </p>
        <div
          class="calc-data-input"
        >
          <h1>
            Введите данные
          </h1>
          <div
            class="calc-input"
          >
            <input
              class="null"
              placeholder="Введите число"
              type="text"
              value=""
            />
          </div>
          <div
            class="calc-input"
          >
            <input
              class="null"
              placeholder="Введите число"
              type="text"
              value=""
            />
          </div>
          <div
            class="add-item"
            tabindex="0"
          >
            +
          </div>
          <button>
            Далее
          </button>
        </div>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
